pkg/stub/handler.go

```diff
@@ -105,145 +103,201 @@ type Handler struct {
 
 	imagestreamFile map[string]string
 	templateFile    map[string]string
-
-	deleteInProgress bool
-
-	mutex *sync.Mutex
 }
 
-func (h *Handler) prepWatchEvent(kind, name string, annotations map[string]string) (*v1alpha1.SamplesResource, string, bool) {
-	if h.deleteInProgress {
-		return nil, "", false
-	}
+// prepWatchEvent decides whether an upsert of the sample should be done, as well as data for either doing the upsert or checking the status of a prior upsert;
+// the return values:
+// - srcfg: return this if we want to check the status of a prior upsert
+// - filePath: used to the caller to look up the image content for the upsert
+// - doUpsert: whether to do the upsert of not ... not doing the upsert optionally triggers the need for checking status of prior upsert
+// - err: if a problem occurred getting the samplesresource, we return the error to bubble up and initiate a retry
+func (h *Handler) prepWatchEvent(kind, name string, annotations map[string]string) (*v1alpha1.SamplesResource, string, bool, error) {
 	srcfg, err := h.sdkwrapper.Get(v1alpha1.SamplesResourceName)
 	if srcfg == nil || err != nil {
-		logrus.Printf("Received watch event %s but ignoring since not have the SamplesResource yet: %#v %#v", kind+"/"+name, err, srcfg)
-		return nil, "", false
+		logrus.Printf("Received watch event %s but not upserting since not have the SamplesResource yet: %#v %#v", kind+"/"+name, err, srcfg)
+		return nil, "", false, err
 	}
+
+	if srcfg.ConditionFalse(v1alpha1.ImageChangesInProgress) {
+		// we do no return the srcfg in these cases because we do not want to bother with any progress tracking
+
+		if srcfg.DeletionTimestamp != nil {
+			logrus.Printf("Received watch event %s but not upserting since deletion of the SamplesResource is in progress and image changes are not in progress", kind+"/"+name)
+			return nil, "", false, nil
+		}
+		switch srcfg.Spec.ManagementState {
+		case operatorsv1alpha1api.Removed:
+			logrus.Debugf("Not upserting %s/%s event because operator is in removed state and image changes are not in progress", kind, name)
+			return nil, "", false, nil
+		case operatorsv1alpha1api.Unmanaged:
+			logrus.Debugf("Not upserting %s/%s event because operator is in unmanaged state and image changes are not in progress", kind, name)
+			return nil, "", false, nil
+		}
+	}
+
 	if annotations != nil {
 		isv, ok := annotations[v1alpha1.SamplesVersionAnnotation]
 		if ok && isv == v1alpha1.CodeLevel {
-			logrus.Debugf("ignoring %s/%s cause operator version matches", kind, name)
+			logrus.Debugf("Not upserting %s/%s cause operator version matches", kind, name)
```

Comment:
followup:  doesn't this mean i can modify the imagestream in any way i want, as long as i don't touch the SamplesVersionAnnotation?

Reply:
Yes that is correct.  This goes back to when I originally asked about checking for this, and asked if we wanted to compare the Spec structs, etc.  You said `no` to that, and suggested we employ the annotation.